chain/chain/src/store/mod.rs

```diff
@@ -2283,12 +2293,13 @@ impl<'a> ChainStoreUpdate<'a> {
             }
             let mut header_hashes_by_height: HashMap<BlockHeight, HashSet<CryptoHash>> =
                 HashMap::new();
+            let mut sync_hash_tracker = crate::state_sync::SyncHashTracker::new();
```

Comment:
1) I see that you're doing the calculations inside of the update. Is that because you want to ensure the consistency between the header processing and the state sync hashes in the db? 
2) Judging by the variable above there may be multiple headers per height. I don't think the tracker deals with that correctly. I guess this is coming back to handling only final blocks - but if you prefer feel free to follow up separately. 
3) Perhaps rather than introducing the tracker, you can adjust the `header_hashes_by_height` to be a BTreeMap that's sorted by heights and just process the headers one by one?

Reply:
1. Yeah so looking at the callers of this function it looks like in practice there's only ever one header in here except for in epoch sync where a few do get added before `commit()` is called. And in that epoch sync case we won't end up storing anything anyway since they're just individual headers without the rest of the headers from that epoch. But in any case the `ChainStoreUpdate::save_block_header()` API doesn't seem to really ask the caller to make sure of anything. But now when calculating this sync hash from these headers, since we do so by looking up the prev header's num_new_chunks and adding to that, we actually do need a requirement that we process the lower heights first. So for example, if someone made a change to `sync_block_headers()` where we save a bunch of headers not necessarily in order of height in one `ChainStoreUpdate` before commiting it, we would need to wait until the end to process them in order

2. I think it will work still, right? Assuming that all the headers added in this `ChainStoreUpdate` have prev headers either stored on disk or in memory as another header in this `ChainStoreUpdate`, if we process them in order of height then our only requirment is that we've previously processed its prev header. This will still be true even if there are multiple of the same height

3. yea sure actually might as well only have one structure here since it's pretty similar. Also while we're here this unqualified `unwrap_or_default()` on `get_all_header_hashes_by_height()` should prob be changed since we only really want to do that if it's `DBNotFound` (which again goes to your other comment in the first PR about how that's kind of ugly and it should maybe just return an Option anyway)

Reply:
1) I'm mostly curious why do it here in the update rather than at a higher level like in block post process. 

2) If there are two headers at the same height, or even if we just process a header that was on a non-canonical fork, the node would still update the new chunks and the state sync hash. This is non-deterministic and different nodes may arrive at different sync hashes. I guess it's not the end of the world but if we can we should try to limit that possibility. Again, all those troubles go away if we delay the processing until final block and again feel free to follow up separately (if it makes sense).

Reply:
1. Oh I see, yeah so in the first version of this PR it was in block post process, but I realized that is actually a bug because we need to do it when saving headers (when you state sync because youre behind you have the headers but not the blocks). But right now we just have this `ChainStoreUpdate::save_block_header()` (and also `save_block_header_no_update_tree()`) that saves it here and is called in a few different places. So this kind of felt like the most natural place to put it, where we can take care of the logic in one place without having to update the callers at all. Let me know if you think of something different though
2. Oh yea true I think updating the sync hash for a non canonical fork is def a bug and should be fixed. Prob by just doing what you say and delaying this until it's actually final